src/main.cpp

```diff
@@ -5170,6 +5170,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             return true;
         }
 
+        if (mapBlockIndex.find(headers[0].hashPrevBlock) == mapBlockIndex.end()) {
+            // Doesn't connect, instead of DoSing in AcceptBlockHeader, request deeper headers
+            if (!IsInitialBlockDownload())
+                pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexBestHeader), uint256());
```

Comment:
Differing philosophy, I guess...Its certainly trivial to fix for the CMPCTBLOCK handling, indeed, I just don't like the idea that our peer might notify us that they have some blocks available and we will simply ignore them because they aren't keeping track of our chainstate. Can we use hashLastUnknownBlock to track if we hit the infinite-loop condition here?